### PR TITLE
Refactor/timelock impl

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:
Extract the implementation of the TimelLock server.

**Implementation Description (bullets)**:
Tried to keep everything that's not implementation specific in `timelock-server`.

**Concerns (what feedback would you like?)**:
General comments on where things were relocated, have I missed something, should I put something back? I also moved some things out of the paxos package because it just didn't seem like the right place, but am open to reconsidering. I had to slightly change `TimelockServerConfiguration`, so that it is agnostic of atomix and paxos. I think these changes are warranted, but would appreciate a second opinion.

**Where should we start reviewing?**:
Anywhere

**Priority (whenever / two weeks / yesterday)**:
We would like to get this in this week

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2076)

### CODE REVIEW

File: docs/source/release_notes/release-notes.rst [hunk 1]
```diff
@@ -42,6 +42,15 @@ develop
...
          - Change
 
+    *    - |devbreak|
+         - The implementation of the TimeLock server has been separated into a new project, `timelock-impl`.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/????>`__)
+
+    *    - |userbreak|
+         - `TimelockServerConfiguration` now requires an algorithm to be specified.
+           Previously, if no algorithm was specified, the configuration would default to atomix.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/????>`__)
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: note: remember to update this before merge

File: docs/source/release_notes/release-notes.rst [hunk 2]
```diff
@@ -42,6 +42,15 @@ develop
     *    - Type
          - Change
 
+    *    - |devbreak|
+         - The implementation of the TimeLock server has been separated into a new project, `timelock-impl`.
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: We should mention that most users should not be directly affected, unless they have been depending on classes from within TimeLock.

File: docs/source/release_notes/release-notes.rst [hunk 3]
```diff
@@ -42,6 +42,15 @@ develop
...
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/????>`__)
+
+    *    - |userbreak|
+         - `TimelockServerConfiguration` now requires an algorithm to be specified.
+           Previously, if no algorithm was specified, the configuration would default to atomix.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/????>`__)
+
```
Comment (jeremyk-91) [inline docs/source/release_notes/release-notes.rst]: Also: this technically causes a user break for consumers of TimeLock metrics (because of the package name change).

File: timelock-impl/src/main/java/com/palantir/atlasdb/timelock/atomix/AtomixRetryer.java [hunk 4]
```diff
@@ -34,7 +34,7 @@
 
 import io.atomix.copycat.session.ClosedSessionException;
 
-public final class AtomixRetryer {
+public final class  AtomixRetryer {
```
Comment (jeremyk-91) [inline timelock-impl/src/main/java/com/palantir/atlasdb/timelock/atomix/AtomixRetryer.java]: I think this was accidentally added.

File: docs/source/release_notes/release-notes.rst (hunk 2 above)
Comment (tpetracca) [inline docs/source/release_notes/release-notes.rst]: Yea at this point I would not consider this a devbreak and just a change in devex

File: timelock-impl/build.gradle [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+jacocoTestReport {
+    executionData test, integTest
+}
+
+check.dependsOn integTest
+
+dependencies {
+    compile project(":timestamp-impl")
+    compile project(":timelock-server")
```
Comment (tpetracca) [inline timelock-impl/build.gradle]: I would expect server to depend on impl, not the other way around.

File: timelock-impl/build.gradle [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    compile group: 'com.palantir.tritium', name: 'tritium-lib'
+    compile group: 'io.atomix', name: 'atomix'
+    compile group: 'io.atomix.catalyst', name: 'catalyst-netty'
+    compile group: 'io.dropwizard', name: 'dropwizard-core', version: libVersions.timelock_dropwizard
+    compile group: 'io.dropwizard', name: 'dropwizard-http2', version: libVersions.timelock_dropwizard
+    compile group: 'io.dropwizard', name: 'dropwizard-request-logging', version: libVersions.timelock_dropwizard
+    compile group: 'io.dropwizard.modules', name: 'dropwizard-java8'
+
+    processor group: 'org.immutables', name: 'value'
+    processor "com.google.auto.service:auto-service:1.0-rc2"
```
Comment (tpetracca) [inline timelock-impl/build.gradle]: Why the change in format and lack of version.props use?

File: timelock-impl/build.gradle [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    testCompile project(":atlasdb-config")
+    testCompile project(path: ":leader-election-impl", configuration: "testArtifacts")
+
+    testCompile group: 'org.assertj', name: 'assertj-core'
+    testCompile group: 'org.mockito', name: 'mockito-core'
+    testCompile project(":atlasdb-tests-shared")
```
Comment (tpetracca) [inline timelock-impl/build.gradle]: nit: block this with other testCompile project stuff.

File: timelock-server-distribution/build.gradle [hunk 8]
```diff
@@ -6,7 +6,7 @@ apply from: "../gradle/shared.gradle"
 apply from: "../gradle/timelock.gradle"
 
 dependencies {
-    runtime project(':timelock-server')
+    runtime project(':timelock-impl')
```
Comment (tpetracca) [inline timelock-server-distribution/build.gradle]: confused why this happened

**Comment by tpetracca** [comment]
So I was expecting the following:
- `timelock-server-distribution` - An sls re-packaging of `timelock-server`
- `timelock-server` - A web server (in our case DropWizard) that mounts resources, filters, etc. provided by `timelock-impl`
- `timelock-impl` - the class(es) that implement or expose the implementations of the various apis: `TimestampService`, `TimestampManagementService`, and `RemoteLockService`. So essentially the `TimeLockResource` class and everything it needs to run.

It looks like the separation here was core/generic timelock impl stuff (which are still in `timelock-server` in this PR) and the actual implementations of the various algorithims (`paxos` and `atomix`, which are now in `timelock-impl`). I don't think separating those things out is a bad thing, but if we did they would probably be separate projects like `timelock-algorithm-paxos` and `timelock-algorithm-atomix`.

**Comment by gmaretic** [comment]
Done a different refactor
